### DISCUSSION TITLE
[SOLVED] Samd51 "Usb Device Not Recognised" ... getting multiple 'Bus Reset'?

### DISCUSSION THREAD

**Discussion by CraigHutchinson** [hathach/tinyusb#921] [Q&A] [answered]
Firstly, I didn't want to report a bug as this may be an issue from my build configuration somehow so help greatly appreciated as appears that newer TinyUSB source code does not work on my Samd51 setup :( 

# Platform
- custom `Samd51J19a` board 'Similar' to `Adafruit Express M4`
- Runtime DFU only to diagnose the issue
- Windows 10 Host OS
- GCC Compiler

# My Issue
On Windows 10 I get a popup and yellow triangle with "Usb Device Not Recognised" 

# When my issue occured
- All was working fine Pre v0.8.0 TinyUsb https://github.com/hathach/tinyusb/commit/589dfdb0e58fe6630f2b67c49c7f682ac51aba8f
- Updating to v0.8.0 or newer tag upto 0.10.1 causes "Usb Device Not Recognised" 

# Log output
This log was generated from the current `master`
Q) The SWO log is showing lots of `Bus Reset` occuring?
```
Hello world from SWO
CDC init

DFU-RUNTIME init

USBD Bus Reset : Full Speed

USBD Setup Received 80 06 00 01 00 00 40 00   Get Descriptor  Queue EP 80 with 18 bytes ... USBD Xfer Complete on EP 80 with 18 bytes

  Queue EP 00 with 0 bytes ... USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 00 05 19 00 00 00 00 00   Set AddressUSBD Xfer Complete on EP 80 with 0 bytes

USBD Bus Reset : Full Speed

USBD Setup Received 80 06 00 01 00 00 40 00   Get Descriptor  Queue EP 80 with 18 bytes ... USBD Xfer Complete on EP 80 with 18 bytes

  Queue EP 00 with 0 bytes ... USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 00 05 1A 00 00 00 00 00   Set AddressUSBD Xfer Complete on EP 80 with 0 bytes

USBD Bus Reset : Full Speed

USBD Setup Received 80 06 00 01 00 00 40 00   Get Descriptor  Queue EP 80 with 18 bytes ... USBD Xfer Complete on EP 80 with 18 bytes

  Queue EP 00 with 0 bytes ... USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 00 05 1B 00 00 00 00 00   Set AddressUSBD Xfer Complete on EP 80 with 0 bytes

USBD Bus Reset : Full Speed

USBD Setup Received 80 06 00 01 00 00 40 00   Get Descriptor  Queue EP 80 with 18 bytes ... USBD Xfer Complete on EP 80 with 18 bytes

  Queue EP 00 with 0 bytes ... USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 00 05 1C 00 00 00 00 00   Set AddressUSBD Xfer Complete on EP 80 with 0 bytes

USBD Suspend 
```

# Diagnosis

I have tried a `git bisect` to no avail, I have gone through lots of the commits to track down when things were broken and identify it is within this range of commits. As mentioned it worked at 589dfdb0e58fe6630f2b67c49c7f682ac51aba8f but not after!
https://github.com/hathach/tinyusb/compare/589dfdb0e58fe6630f2b67c49c7f682ac51aba8f...0.8.0

Notably I 'think' its since "merge class driver control_request & control_complete to control_xfer…"https://github.com/hathach/tinyusb/commit/3c31d0805130e9551f97afae94f765b5d032991b

# Theory
Maybe the newer code isn't tolerant to tud_task() not being called at super regular intervals? I just fixed a sandbox test by reducing a sleep-loop duration and it now does work with the latest code... umm

**Comment by CraigHutchinson** [comment, marked as answer]
I finally tracked it down and was an issue with `tud_vendor_control_request_cb()` that has changed, managed to ommit the fact I was using a BOS descriptor in the OP but changing his fixes the issue I was facing:

```
if TUSB_VERSION_MINOR < 8
bool tud_vendor_control_request_cb( uint8_t rhport, tusb_control_request_t const* request )
{
#else
extern "C" bool tud_vendor_control_xfer_cb(uint8_t rhport, uint8_t stage, tusb_control_request_t const* request)
{
    // nothing to with DATA & ACK stage
    if (stage != CONTROL_STAGE_SETUP) return true;
    if (request->bmRequestType_bit.type!= TUSB_REQ_TYPE_VENDOR) return true; // stall unknown request
#endif
```

**Reply by hathach** [reply]
glad you could work it out :+1:

**Reply by neilh10** [reply]
Hey thanks for sharing, I have a project I want to do with the SAMD51.

**Reply by CraigHutchinson** [reply]
@hathach Could it be possible to 'poison' the old function names? e.g. https://www.fluentcpp.com/2018/09/04/function-poisoning-in-cpp/ I can create a feature request if so as it could help with people moving/upgrading between versions in future ...possibly.
EDIT: i.e. `#pragma GCC poison tud_vendor_control_request_cb` in the `tusb.h` somewhere.

**Reply by hathach** [reply]
@CraigHutchinson that is a great suggestion, though, it seem only gcc support it. But good enough to put it in, here we go https://github.com/hathach/tinyusb/pull/934 . Thanks, I learn something new today :+1:

**Reply by dhalbert** [reply]
There are some suggestions in the comments in https://www.fluentcpp.com/2018/09/04/function-poisoning-in-cpp/ about what to do for other compilers. In MSVC, one suggestion was just to do `#define dont_use_me()`, with the wrong number of arguments. There is also a `deprecated` feature in MSVC. And `clang` also implements `poison`.

**Reply by hathach** [reply]
> `#define dont_use_me()`, with the wrong number of arguments.

This is very clever and could work with all other compiler. Thanks for the tip, will update it for other compiler with this :+1:

**Reply by CraigHutchinson** [reply]
I've not even considered compiling TinyUsb using MSVC... they do have an ARM compiler, is this actually supported by TinyUSB already?

**Reply by hathach** [reply]
I don't think MSVC supported arm mcu. It is just suggestion to work with other compiler. The notable compiler for embedded would be IAR, TI, Rensas etc.. those that aren't gcc based one.
